### PR TITLE
key rollover API doc

### PR DESCRIPTION
**Pull request by yumingcao-okta**
This is the API doc for EA Key rollover feature with Postman collection.

Changes are made under `/apps` section.

@karlmcguinness-okta @tthompson-okta @joelfranusic-okta @bng-okta Please review.

### CODE REVIEW

File: _source/docs/api/resources/apps.md [hunk 1]
```diff
@@ -260,6 +263,7 @@ Specifies credentials and scheme for the application's `signOnMode`.
 | ---------------- | -------------------------------------------------------------------------------------------------------------- | ----------------------------------------------------- | -------- | --------------- | --------- | --------- | ---------- |
 | scheme           | Determines how credentials are managed for the `signOnMode`                                                    | [Authentication Scheme](#authentication-schemes)      | TRUE     |                 |           |           |            |
 | userNameTemplate | Template used to generate a user’s username when the application is assigned via a group or directly to a user | [UserName Template Object](#username-template-object) | TRUE     | *Okta UserName* |           |           |            |
+| signing          | Contains app certificate information                                                                           | [Signing Object](#signing-object)                     | TRUE     |                 |           |           |            |
```
Comment (karlmcguinness-okta) [inline _source/docs/api/resources/apps.md]: description: Signing credential for the `signOnMode`
type: Signing Credential Object

Is this object null for SWA modes?

File: _source/docs/api/resources/apps.md [hunk 2]
```diff
@@ -317,6 +324,25 @@ Specifies the template used to generate a user's username when the application i
 }
 ~~~
 
+#### Signing Object
+Shows id of app certificate that is used for SSO signing and SAML assertion validation
```
Comment (karlmcguinness-okta) [inline _source/docs/api/resources/apps.md]: Signing Credential Object

Determines the key used for signing assertions for the `signOnMode`

File: _source/docs/api/resources/apps.md [hunk 3]
```diff
@@ -317,6 +324,25 @@ Specifies the template used to generate a user's username when the application i
...
+
+|------------+-----------------------------------------+----------------------------------+----------+-------------------+-----------+-----------+------------|
+| Property   | Description                             | DataType                         | Nullable | Default           | MinLength | MaxLength | Validation |
+| ---------- | --------------------------------------- | -------------------------------- | -------- | ----------------- | --------- | ----------| ---------- |
+| kid        | app certificate id                      | String                           | TRUE     |                   |           |           |            |
```
Comment (karlmcguinness-okta) [inline _source/docs/api/resources/apps.md]: When is the property `null`?

Description: Key reference for a non-expired key in the [app's key store](#application-key-store)

File: _source/docs/api/resources/apps.md [hunk 4]
```diff
@@ -3368,3 +3394,475 @@ curl -v -X DELETE \
 ~~~json
 {}
 ~~~
+
+## Application Certificate Model
```
Comment (karlmcguinness-okta) [inline _source/docs/api/resources/apps.md]: Application Key Model

File: _source/docs/api/resources/apps.md [hunk 5]
```diff
@@ -3368,3 +3394,475 @@ curl -v -X DELETE \
...
+
+The application certificate model represents an application certificate in [JSON Web Key](https://tools.ietf.org/html/rfc7517) format.
```
Comment (karlmcguinness-okta) [inline _source/docs/api/resources/apps.md]: The application key credential model defines a [JSON Web Key](https://tools.ietf.org/html/rfc7517) for a signature or encryption credential for an application.

> Currently only the X.509 JWK format is supported for applications with the `SAML_2_0` sign-on mode.

File: _source/docs/api/resources/apps.md [hunk 6]
```diff
@@ -3368,3 +3394,475 @@ curl -v -X DELETE \
...
+  "use": "sig"
+}
+~~~
+
+### Application Certificate Properties
+
+|------------------+--------------------------------------------------------------+-----------------------------------------------------------------------------|----------|--------|----------|-----------|-----------+------------|
+| Property         | Description                                                  | DataType                                                                    | Nullable | Unique | Readonly | MinLength | MaxLength | Validation |
+| ---------------- | ------------------------------------------------------------ | --------------------------------------------------------------------------- | -------- | ------ | -------- | --------- | --------- | ---------- |
+| created          | timestamp when app certificate was created                   | Date                                                                        | FALSE    | FALSE  | TRUE     |           |           |            |
```
Comment (karlmcguinness-okta) [inline _source/docs/api/resources/apps.md]: timestamp when certificate was created

File: _source/docs/api/resources/apps.md [hunk 7]
```diff
@@ -3368,3 +3394,475 @@ curl -v -X DELETE \
...
+| expiresAt        | timestamp when app certificate expires                       | Date                                                                        | FALSE    | FALSE  | TRUE     |           |           |            |
```
Comment (karlmcguinness-okta) [inline _source/docs/api/resources/apps.md]: timestamp when certificate expires

File: _source/docs/api/resources/apps.md [hunk 8]
```diff
@@ -3368,3 +3394,475 @@ curl -v -X DELETE \
...
+| x5c              | X.509 certificate chain                                      | Array                                                                       | TRUE     | TRUE   | TRUE     |           |           |            |
+| x5t#S256         | X.509 certificate SHA-256 thumbprint                         | String                                                                      | TRUE     | TRUE   | TRUE     |           |           |            |
+| kid              | certificate id                                               | String                                                                      | FALSE    | TRUE   | TRUE     |           |           |            |
```
Comment (karlmcguinness-okta) [inline _source/docs/api/resources/apps.md]: unique identifier for the certificate (key)

File: _source/docs/api/resources/apps.md [hunk 9]
```diff
@@ -3368,3 +3394,475 @@ curl -v -X DELETE \
...
+| kty              | cryptographic algorithm family used with the certificate     | String                                                                      | FALSE    | FALSE  | TRUE     |           |           |            |
```
Comment (karlmcguinness-okta) [inline _source/docs/api/resources/apps.md]: cryptographic algorithm family for the certificate's keypair

File: _source/docs/api/resources/apps.md [hunk 10]
```diff
@@ -3368,3 +3394,475 @@ curl -v -X DELETE \
...
+| use              | usage of the certificate                                     | String                                                                      | TRUE     | FALSE  | TRUE     |           |           |            |
```
Comment (karlmcguinness-okta) [inline _source/docs/api/resources/apps.md]: acceptable usage of the certificate

File: _source/docs/api/resources/apps.md [hunk 11]
```diff
@@ -3368,3 +3394,475 @@ curl -v -X DELETE \
...
+|------------------+--------------------------------------------------------------+-----------------------------------------------------------------------------|----------|--------|----------|-----------|-----------+------------|
+
+## Application Certificate Operations
```
Comment (karlmcguinness-okta) [inline _source/docs/api/resources/apps.md]: Application Key Store Operations

File: _source/docs/api/resources/apps.md [hunk 12]
```diff
@@ -3368,3 +3394,475 @@ curl -v -X DELETE \
...
+
+> You need to enable `KEY_ROLLOVER` feature to perform the following operations, it is advised to contact Customer Support before enabling this feature.
```
Comment (karlmcguinness-okta) [inline _source/docs/api/resources/apps.md]: Generate New Application Key Credential

File: _source/docs/api/resources/apps.md [hunk 13]
```diff
@@ -3368,3 +3394,475 @@ curl -v -X DELETE \
...
+
+### Generate Certificate for Application
+{:.api .api-operation}
+
+<span class="api-uri-template api-uri-post"><span class="api-label">POST</span> /apps/*:aid*/credentials/keys/generate</span>
+
+Generate certificate for an application
```
Comment (karlmcguinness-okta) [inline _source/docs/api/resources/apps.md]: Generate's a new X.509 certificate for an application key credential

File: _source/docs/api/resources/apps.md [hunk 14]
```diff
@@ -3368,3 +3394,475 @@ curl -v -X DELETE \
...
+
+Parameter     | Description                                     | Param Type | DataType                                      | Required | Default
+------------- | ----------------------------------------------- | ---------- | --------------------------------------------- | -------- | -------
+aid           | unique key of [Application](#application-model) | URL        | String                                        | TRUE     |
+validityYears | validity years since certificate creation date  | Query      | Number                                        | TRUE     |
+
+##### Response Parameters
+{:.api .api-response .api-response-params}
+
+Return the generated [Application Certificate](#application-certificate-model).
```
Comment (karlmcguinness-okta) [inline _source/docs/api/resources/apps.md]: Application Key Credential

File: _source/docs/api/resources/apps.md [hunk 15]
```diff
@@ -3368,3 +3394,475 @@ curl -v -X DELETE \
...
+
+Generate certificate for an application
+
+##### Request Parameters
+{:.api .api-request .api-request-params}
+
+Parameter     | Description                                     | Param Type | DataType                                      | Required | Default
+------------- | ----------------------------------------------- | ---------- | --------------------------------------------- | -------- | -------
+aid           | unique key of [Application](#application-model) | URL        | String                                        | TRUE     |
+validityYears | validity years since certificate creation date  | Query      | Number                                        | TRUE     |
```
Comment (karlmcguinness-okta) [inline _source/docs/api/resources/apps.md]: expiry of the [application key credential](#application-key-credential-model)

File: _source/docs/api/resources/apps.md [hunk 16]
```diff
@@ -3368,3 +3394,475 @@ curl -v -X DELETE \
...
+    ],
+    "kid": "akm5hvbn1vojA9Fsa0h7",
+    "kty": "RSA",
+    "use": "sig",
+    "x5t#S256": "7CCyXWwKzH4P6PoBP91B1S_iIZVzuGffVnUXu-BTYQQ"
+  }
+]
+~~~
+
+### Get Certificate for Application
```
Comment (karlmcguinness-okta) [inline _source/docs/api/resources/apps.md]: Get Key for Application

File: _source/docs/api/resources/apps.md [hunk 17]
```diff
@@ -3368,3 +3394,475 @@ curl -v -X DELETE \
...
+  "errorId": "oaeMHrsk2WLTACvPU5T7yQ4yw",
+  "errorCauses": [
+    {
+      "errorSummary": "Validity years out of range. It should be 2 - 10 years"
+    }
+  ]
+}
+~~~
+
+### List Certificates for Application
```
Comment (karlmcguinness-okta) [inline _source/docs/api/resources/apps.md]: List Keys for Application

File: _source/docs/api/resources/apps.md [hunk 18]
```diff
@@ -3368,3 +3394,475 @@ curl -v -X DELETE \
...
+{:.api .api-operation}
+
+<span class="api-uri-template api-uri-get"><span class="api-label">GET</span> /apps/*:aid*/credentials/keys</span>
+
+Enumerates certificates for an Application
```
Comment (karlmcguinness-okta) [inline _source/docs/api/resources/apps.md]: Enumerates key credentials for an Application

File: _source/docs/api/resources/apps.md [hunk 19]
```diff
@@ -3368,3 +3394,475 @@ curl -v -X DELETE \
...
+
+Get a specific certificate for an application by `kid`
+
+##### Request Parameters
+{:.api .api-request .api-request-params}
+
+Parameter     | Description                                                             | Param Type | DataType                                      | Required | Default
+------------- | ----------------------------------------------------------------------- | ---------- | --------------------------------------------- | -------- | -------
+aid           | unique key of [Application](#application-model)                         | URL        | String                                        | TRUE     |
+kid           | unique key of [Application Certificate](#application-certificate-model) | URL        | String                                        | TRUE     |
```
Comment (karlmcguinness-okta) [inline _source/docs/api/resources/apps.md]: Update reference to new name

File: _source/docs/api/resources/apps.md [hunk 20]
```diff
@@ -3368,3 +3394,475 @@ curl -v -X DELETE \
...
+  }
+]
+~~~
+
+### Get Certificate for Application
+{:.api .api-operation}
+
+<span class="api-uri-template api-uri-get"><span class="api-label">GET</span> /apps/*:aid*/credentials/keys/*:kid*</span>
+
+Get a specific certificate for an application by `kid`
```
Comment (karlmcguinness-okta) [inline _source/docs/api/resources/apps.md]: Gets a specific [application key credential](#application-key-credential-model) by `kid`

File: _source/docs/api/resources/apps.md [hunk 21]
```diff
@@ -3368,3 +3394,475 @@ curl -v -X DELETE \
...
+  ],
+  "kid": "akm5hvbbevE341ovl0h7",
+  "kty": "RSA",
+  "use": "sig",
+  "x5t#S256": "5GOpy9CQVtfvBmu2T8BHvpKE4OGtC3BuS046t7p9pps"
+}
+~~~
+
+### Update Certificate for Application
```
Comment (karlmcguinness-okta) [inline _source/docs/api/resources/apps.md]: Update Key for Application

File: _source/docs/api/resources/apps.md [hunk 22]
```diff
@@ -3368,3 +3394,475 @@ curl -v -X DELETE \
...
+
+SAML metadata in XML
+
+##### Request Example
+{:.api .api-request .api-request-example}
+
+~~~sh
+curl -v -X GET \
+-H "Accept: application/json" \
+-H "Content-Type: application/json" \
```
Comment (karlmcguinness-okta) [inline _source/docs/api/resources/apps.md]: This needs to be `application/xml`

**Comment by karlmcguinness-okta** [comment]
Reviewed with comments, just some renames of model names and operations.  Thx for doing this!
On Dec 14, 2015, at 3:38 PM, Yuming Cao <notifications@github.com<mailto:notifications@github.com>> wrote:


This is the API doc for EA Key rollover feature with Postman collection.

Changes are made under /apps section.

@karlmcguinness-okta<https://github.com/karlmcguinness-okta> @tthompson-okta<https://github.com/tthompson-okta> @joelfranusic-okta<https://github.com/joelfranusic-okta> @bng-okta<https://github.com/bng-okta> Please review.

________________________________
You can view, comment on, or merge this pull request online at:

  https://github.com/okta/okta.github.io/pull/117

Commit Summary

  *   doc key rollover endpoints
  *   add collection

File Changes

  *   M _source/docs/api/postman/apps.json<https://github.com/okta/okta.github.io/pull/117/files#diff-0> (642)
  *   M _source/docs/api/resources/apps.md<https://github.com/okta/okta.github.io/pull/117/files#diff-1> (498)

Patch Links:

  *   https://github.com/okta/okta.github.io/pull/117.patch
  *   https://github.com/okta/okta.github.io/pull/117.diff

File: _source/docs/api/resources/apps.md (hunk 1 above)
Comment (yumingcao-okta) [inline _source/docs/api/resources/apps.md]: For now this object is not null when it's SWA modes, please let me know if you want to change this behavior.

File: _source/docs/api/resources/apps.md (hunk 3 above)
Comment (yumingcao-okta) [inline _source/docs/api/resources/apps.md]: Thanks for catching this, the property is an empty json object (`{}`) when feature is disabled, but still not `null`. Will change to `FALSE`.

For now, `kid` is displayed no matter the key is expired or not, let me know if you only want to display non-expired key here.

File: _source/docs/api/resources/apps.md (hunk 5 above)
Comment (yumingcao-okta) [inline _source/docs/api/resources/apps.md]: My understanding here is signature credential is used for signing SAML assertions, while encryption credential is used for validating the signed SAML assertions, it that correct?

File: _source/docs/api/resources/apps.md (hunk 12 above)
Comment (yumingcao-okta) [inline _source/docs/api/resources/apps.md]: Shall we say `Generate New Application Key Credential` here?

File: _source/docs/api/resources/apps.md (hunk 17 above)
Comment (yumingcao-okta) [inline _source/docs/api/resources/apps.md]: List Key Credentials for Application?

File: _source/docs/api/resources/apps.md (hunk 16 above)
Comment (yumingcao-okta) [inline _source/docs/api/resources/apps.md]: Get Key Credential for Application?

File: _source/docs/api/resources/apps.md (hunk 22 above)
Comment (yumingcao-okta) [inline _source/docs/api/resources/apps.md]: Nice catch, thanks!

**Comment by yumingcao-okta** [comment]
@karlmcguinness-okta Thanks for looking into this! Replied some comments with questions.

File: _source/docs/api/resources/apps.md (hunk 1 above)
Comment (karlmcguinness-okta) [inline _source/docs/api/resources/apps.md]: I think we should only expose the `signing` property for sign on modes that support key rollover (e.g. SAML)

File: _source/docs/api/resources/apps.md (hunk 3 above)
Comment (karlmcguinness-okta) [inline _source/docs/api/resources/apps.md]: We should hide the property when feature flag is disabled.  The kid should always be display independent of whether its expired or not.  Expiration is an attribute of the key but we should keep the reference.

File: _source/docs/api/resources/apps.md (hunk 5 above)
Comment (karlmcguinness-okta) [inline _source/docs/api/resources/apps.md]: I was a little ahead of myself.  You are correct that the key today is only for signatures.  I was trying to keep the model generic but the use case specific (e.g. `signing` property).  In the future we will add an `encrytion` property and allow customers to upload public keys.

File: _source/docs/api/resources/apps.md (hunk 12 above)
Comment (karlmcguinness-okta) [inline _source/docs/api/resources/apps.md]: sure

File: _source/docs/api/resources/apps.md (hunk 3 above)
Comment (yumingcao-okta) [inline _source/docs/api/resources/apps.md]: ack, and here's new description proposal: Reference for key credential in the [app's key store](#application-key-store-operations)

File: _source/docs/api/resources/apps.md (hunk 1 above)
Comment (yumingcao-okta) [inline _source/docs/api/resources/apps.md]: ack

File: _source/docs/api/resources/apps.md (hunk 5 above)
Comment (yumingcao-okta) [inline _source/docs/api/resources/apps.md]: Got it.

**Comment by yumingcao-okta** [comment]
@karlmcguinness-okta Changes made.

**Comment by yumingcao-okta** [comment]
@KarlMcGuinness 2nd round comments addressed (key model section moved up under app group model).

File: _source/docs/api/resources/apps.md [hunk 23]
```diff
@@ -3368,3 +3573,259 @@ curl -v -X DELETE \
...
+NdkQVwSmWMh6KLxVM18/xAkq8zyKbMbvQnTjFB7x45bgokwbjhivWqrB5LYHHCVN7k/8mKlS4eCK
+Ci6RGEmErjojr4QN2xV0qAqP6CcGANgpepsQJCzlWucMFKAh0x9Kl8fmiQodfyLXyrebYsVnLrMf
+jxE1b6dg4jKvv975tf5wreQSYZ7m//g3/+NnuDKkN/03HqhV7hTNi1fyctXk8I5Nwgyr+pT5LT2k
+YoEdncuy+GQGzE9yLOhC4HNfHQXpqp2tMPdRlw==</ds:X509Certificate>
+                </ds:X509Data>
+            </ds:KeyInfo>
+        </md:KeyDescriptor>
+        <md:NameIDFormat>urn:oasis:names:tc:SAML:1.1:nameid-format:emailAddress</md:NameIDFormat>
+        <md:NameIDFormat>urn:oasis:names:tc:SAML:1.1:nameid-format:unspecified</md:NameIDFormat>
+        <md:SingleSignOnService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST" Location="https://balacomptest.oktapreview.com/app/salesforce-fedid/exk39sivhuytV2D8H0h7/sso/saml"/>
```
Comment (bng-okta) [inline _source/docs/api/resources/apps.md]: should we update `balacomptest` to `exampleorg` in the sample response? What do we do for sample responses in the other parts of the doc?

and also `salesforce-fedid` to `sample-app` or that might be okay.

Comment (yumingcao-okta) [inline _source/docs/api/resources/apps.md]: This will be addressed in next PR. Thanks @bng-okta!

**Pull request merged**